File: Judges/UVA/Training/665.cpp
#include <iostream>
#include <bitset>
#include <vector>

using namespace std;

string TRASH;
int __INT_VAL, __CHAR_VAL;
#define cinInt() (cin >> __INT_VAL, __INT_VAL)
#define cinChar() (cin >> __CHAR_VAL, __CHAR_VAL)

#define each(item, list) (vector<int>::iterator item = list.begin(); item != list.end(); item++)
#define range(i, max) (int i = 0; i < max; i++)

#define dbug if (false) cout

const int Light = 0;
const int Heavy = 1;

int main()
{
	int instances = cinInt();
	dbug << "instances: " << instances << endl;

	for (int instance = 0; instance < instances; instance++)
	{
		int nCoins = cinInt();
		int nWeightings = cinInt();
		bitset<100>* coins = new bitset<100>[2]{ bitset<100>(0x0), bitset<100>(0x0) };

		dbug << "isntance: " << instance << endl;
		dbug << "\tnCoins: " << nCoins << endl;
		dbug << "\tnWeightings: " << nWeightings << endl;

		dbug << "\tcoins[Light]:";
		for range(coin, nCoins)
			dbug << " " << coins[Light][coin];
		dbug << endl;

		dbug << "\tcoins[Heavy]:";
		for range(coin, nCoins)
			dbug << " " << coins[Heavy][coin];
		dbug << endl;

		for (int weighting = 0; weighting < nWeightings; weighting++)
		{
			int nCoinsOnSide = cinInt();
			dbug << "\t\tnCoinsOnSide: " << nCoinsOnSide << endl;
			
			vector<int> leftCoins;
			vector<int> rightCoins;

			dbug << "\t\tleftCoins:";
			for range(coin, nCoinsOnSide) {
				leftCoins.push_back(cinInt() - 1);
				dbug << " " << *(leftCoins.rbegin());
			}
			dbug << endl;

			dbug << "\t\trightCoins:";
			for range(coin, nCoinsOnSide) {
				rightCoins.push_back(cinInt() - 1);
				dbug << " " << *(rightCoins.rbegin());
			}
			dbug << endl;

			string signal; cin >> signal;
			dbug << "\t\tsignal: '" << signal << "'" << endl;

			if (signal[0] != '>')
			{
				for each(coin, leftCoins)
					coins[Light][*coin] = true;

				dbug << "\t\t!> coins[Light]:";
				for range(coin, nCoins)
					dbug << " " << coins[Light][coin];
				dbug << endl;

				for each(coin, rightCoins)
					coins[Heavy][*coin] = true;

				dbug << "\t\t!> coins[Heavy]:";
				for range(coin, nCoins)
					dbug << " " << coins[Heavy][coin];
				dbug << endl;
			}
			
			if (signal[0] != '<')
			{
				for each(coin, leftCoins)
					coins[Heavy][*coin] = true;

				dbug << "\t\t!< coins[Heavy]:";
				for range(coin, nCoins)
					dbug << " " << coins[Heavy][coin];
				dbug << endl;

				for each(coin, rightCoins)
					coins[Light][*coin] = true;

				dbug << "\t\t!< coins[Light]:";
				for range(coin, nCoins)
					dbug << " " << coins[Light][coin];
				dbug << endl;
			}

			dbug << endl;
		}

		int ans = 0;

		for (int i = 0; i < 2 && ans != -1; i++)
		{
			for range(coin, nCoins) {
				dbug << "on " << coin << endl;
				if (!coins[i][coin]) {
					if (ans != 0)
					{
						dbug << "for ans " << ans << endl;
						if (ans != coin + 1)
						{
							dbug << "Happened: n = -1" << endl;
							ans = -1;
							break;
						}
					}
					else ans = coin + 1;
				}
			}
		}

		if (ans == -1) ans = 0;

		if (instance > 0) cout << endl;
		cout << ans << endl;
	}
}

File: Judges/UVA/Training/665.py
class struct: pass

cases = input()

for c in xrange(cases):
	raw_input()

	raw_in = raw_input().split()
	n_coins = int(raw_in[0])
	n_weightings = int(raw_in[1])

	# print 'case: ' + str(c)
	# print '\tn_coins: ' + str(n_coins)
	# print '\tn_weightings: ' + str(n_weightings)

	# Vars
	coins = [ False for i in xrange(n_coins) ]
	weightings = []

	for w in xrange(n_weightings):
		raw_in = raw_input().split()

		weighting = struct()
		weighting.n_coins = int(raw_in.pop(0))
		weighting.left = [ int(coin) - 1 for coin in raw_in[: weighting.n_coins] ]
		weighting.right = [ int(coin) - 1 for coin in raw_in[weighting.n_coins :] ]
		weighting.signal = raw_input()

		# print '\t\tn_coins: ' + str(weighting.n_coins)
		# print '\t\tleft: ' + str(weighting.left)
		# print '\t\tright: ' + str(weighting.right)
		# print '\t\tsignal: ' + weighting.signal
		# print ''

		# Processing =
		
		if weighting.signal == '=':
			for coin in weighting.left:
				coins[coin] = True
			for coin in weighting.right:
				coins[coin] = True
		else:
			weightings.append(weighting)

	# Cases where the coins are light and heavy
	coins_cases = [ list(coins), list(coins) ]

	for weighting in weightings:
		if weighting.signal == '<':
			for coin in weighting.left:
				coins_cases[0][coin] = True
			for coin in weighting.right:
				coins_cases[1][coin] = True
		else:
			for coin in weighting.left:
				coins_cases[1][coin] = True
			for coin in weighting.right:
				coins_cases[0][coin] = True

	# print 'coins_cases[0]: ' + str(coins_cases[0])
	# print 'coins_cases[1]: ' + str(coins_cases[1])

	false_coin = [0] * len(coins_cases)

	for i_case in xrange(len(coins_cases)):
		coins_case = coins_cases[i_case]

		for i_coin in xrange(len(coins_case)):
			if not coins_case[i_coin]:
				if false_coin[i_case] == 0:
					false_coin[i_case] = i_coin + 1
				else:
					false_coin = 0
					break

		if false_coin == 0:
			break

	if false_coin != 0:
		if false_coin[0] > 0 and false_coin[1] > 0:
			if false_coin[0] == false_coin[1]:
				false_coin = false_coin[0]
			else:
				false_coin = 0
		else:
			false_coin = false_coin[0] + false_coin[1]
	
	# Break line
	if c > 0: print ''

	print false_coin